src/components/BuySell/components/MaxButton.tsx

```diff
@@ -10,56 +10,50 @@ const MaxButton: React.FC = () => {
     buySellToken,
     isUserBuying,
     selectedCurrency,
-    uniswapData,
     spendingTokenBalance,
-    onSetCurrencyQuantity,
-    onSetTokenQuantity,
     onSetActiveField,
+    zeroExTradeData,
+    onSetAmount,
   } = useBuySell()
 
   const { account } = useWallet()
 
   // Do not allow users to spend maximum ETH quantity due to gas cost complications
-  const isMaxSpendDisabled =
-    isUserBuying && selectedCurrency?.id === 'wrapped_eth'
+  const isMaxSpendDisabled = isUserBuying && selectedCurrency?.label === 'ETH'
 
   let spendingTokenSymbol = 'ETH'
   let buttonAction = () => {}
-  let requiredQuantity = new BigNumber(uniswapData?.amount_in || 0).dividedBy(
-    new BigNumber(10).pow(18)
-  )
+  let requiredQuantity = new BigNumber(zeroExTradeData?.maxInput || 0)
 
   if (!isUserBuying) {
     spendingTokenSymbol = buySellToken.toUpperCase()
     buttonAction = () => {
       onSetActiveField('set')
-      onSetTokenQuantity(spendingTokenBalance.toString())
+      onSetAmount(spendingTokenBalance.toString())
     }
-  } else if (selectedCurrency?.id === 'wrapped_eth') {
+  } else if (selectedCurrency?.label === 'ETH') {
     spendingTokenSymbol = selectedCurrency.label
     buttonAction = () => {
       onSetActiveField('currency')
-      onSetCurrencyQuantity(spendingTokenBalance.toString())
+      onSetAmount(spendingTokenBalance.toString())
     }
-  } else if (selectedCurrency?.id === 'mcd') {
+  } else if (selectedCurrency?.label === 'DAI') {
     spendingTokenSymbol = selectedCurrency.label
     buttonAction = () => {
       onSetActiveField('currency')
-      onSetCurrencyQuantity(spendingTokenBalance.toString())
+      onSetAmount(spendingTokenBalance.toString())
     }
-  } else if (selectedCurrency?.id === 'usdc') {
+  } else if (selectedCurrency?.label === 'USDC') {
     spendingTokenSymbol = selectedCurrency.label
     buttonAction = () => {
       onSetActiveField('currency')
-      onSetCurrencyQuantity(spendingTokenBalance.toString())
+      onSetAmount(spendingTokenBalance.toString())
     }
-    requiredQuantity = new BigNumber(uniswapData?.amount_in || 0).dividedBy(
-      new BigNumber(10).pow(6)
-    )
+    requiredQuantity = new BigNumber(zeroExTradeData?.maxInput || 0)
   }
 
   const userHasSufficientFunds = spendingTokenBalance.isGreaterThanOrEqualTo(
-    requiredQuantity
+    requiredQuantity.multipliedBy(0.99999) // accounts for rounding errors
```

Comment:
The root of this problem is that the useBalances hook returns balances in displayable values, rather than their precise (1e18) values. We end up losing some precision when balances get rounded during this conversion. We should definitely, think about refactoring the useBalances hook to fix this in the future, but it is quite deeply integrated into other parts of the codebase, so it probably deserves a separate PR

Reply:
Makes sense, can you write something to that effect in the comment so we have that context in the future? maybe mention we lose precision in `BalancesProvider` and this rounds down to account for that.